[GEOS-11284] Promote community module "datadir catalog loader" to core

groldan opened the pull request:
[![GEOS-11284](https://badgen.net/badge/JIRA/GEOS-11284/0052CC)](https://osgeo-org.atlassian.net/browse/GEOS-11284) [<img width="16" alt="Powered by Pull Request Badge" src="https://user-images.githubusercontent.com/1393946/111216524-d2bb8e00-85d4-11eb-821b-ed4c00989c02.png">](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

Promote the "datadir catalog loader" community module to the GeoServer core,  improving startup performance for deployments with large data directories.

- Moves the `org.geoserver.datadir` package from the community module to `gs-main`
- Removes the `community/datadir-catalog-loader` module
- Adds configuration docs to `doc/en/user/source/datadirectory/setting.rst`
- Fixes failing tests when used as the default geoserver loader due to lack of support for legacy catalogs, and in setting the default workspace.

---
* Supersedes #7421
* Further improved by:
  * #8382
  * #8383
  * #8384
  * #8413 
  * #8414

<!--Include a few sentences describing the overall goals for this Pull Request-->
  
<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

The PR will be merged when all the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)), there is a code committer review, and the checklist has been fulfilled.

groldan commented:
@aaime and others, the old PR was a mess to rebase after the Palantir formatting.
Also, curiously enough, your [sample datadir](https://www.dropbox.com/scl/fi/ip3x82pdzui1rxnvwsqmb/large_shapefiles.tar.bz2?rlkey=7uk600hlcp9z6tdbsdka86dpl&st=bkz2knvy&dl=1), worked with the community module as it was, but not with the refactored code from the old pr.

In any case, I've cleaned things up a lot in this one, making the new loader play nicely with GeoServerLoader and DefaultGeoServerLoader, instead of re-doing a lot of their workflow.

Added a bunch of test cases for invalid datadir configs.

The next commit will be about removing the community module and moving these classes to core.

Meanwhile, any testing is very much welcomed.

You can download these two files to run GeoServer against the sample data directory from docker compose:
* [initialize_volume.sh](https://gist.github.com/groldan/b09df3377304915e58c3cc90de3706df#file-initialize_volume-sh): creates a docker volume named `geoserver_nfs_data`. Downloads the sample datadir file if it's not present in the current directory.
* [compose.yml](https://gist.github.com/groldan/702970e69d18cef763a875ac21fc5e28#file-compose-yml): GeoServer docker composition to load the catalog from an NFS share using geoserver_nfs_data volume

So just download both files and run
```shell
chmod +x ./initialize_volume.sh
./initialize_volume.sh
docker compose up -d
```
When ready GeoServer will be available at http://localhost:8080/geoserver/

groldan commented:
> Currently failing on `app-schema-test` but I've no clue what may be causing it. Looking into it

The app schema tests use a legacy data directory and the new loader was [missing a check for it](https://github.com/geoserver/geoserver/compare/315b23474cf8fbd11ccb6416647ce8d0cdfe38e3..5ae48370869baa26570981089cf08e2588c7fc42).

aaime commented:
I've tested this latest version, it does not lockup anymore. However, at the same time, I don't see a speed-up anymore, when data directories on the local disk, at least. In both cases the dirs are on a SSD and I've dropped the file system caches before running the bench:


Data dir   | Workspaces | Layers | Current loader | New Loader |
--------------|-------------------|-----------|-----------------------|-------------------|
LargeWS | 1000              | 16k       |  120 sec             | 112 sec         |
LargeLayer| 1                  |  1k        | 23 sec                | 23 sec           |

If you consider the current loader is logging every single layer, feature type, store and resource loaded, and the new one doesn't, those 10 seconds in the large case are going to be even less.
By the way, why was the logging modified? I can see it's limiting performance , but one just needs to set up the logging level to "production" and the issue is gone. For smaller data directories having the logs on is useful.

I've also tried the LargeWS on a Samba share (disk attached to the ADSL router), and got the following timings, but in both cases I stopped the startup after 10 minutes, only a small portion of the catalog was loaded by then (I changed the logging of catalog objects being loaded from FINE to CONFIG to see what was happening).

Can you double check there is any speedup left by using only this PR (leaving out the others about encryption and the like)? Maybe the speedup is visible only with a NFS server.

groldan commented:
Hi Andrea,

You're right; as of #ec7a87ac4ec649b62045778ac6ebcf8a7b351666, the performance improvement is not great.
There was a reason for that but it's **fixed** for good now. I'll explain below.

Meanwhile, I've added commit #26c46e4201714fa9625c20645cc3c110ed42e547 to align the logging of the new loader with the current loader (i.e., `config` level log entry for each object added to the catalog, so it’s actually a bit more detailed than the current loader).

# TL;DR

The updated `DataDirectoryGeoServerLoader` (as of this last commit) delivers **faster and safer** catalog loading by parsing in parallel with a work-stealing thread pool and adding sequentially to avoid deadlocks. [GEOS-11758] initially reduced performance gains by prioritizing safety—shifting parsing of Stores, Layers, and LayerGroups to the main thread to prevent Spring-related deadlocks—costing some parallelism. This is now fixed by restoring parallel parsing while ensuring sequential adds for unsafe types, achieving up to 3x faster startup times for large catalogs without compromising reliability.

Data dir (layers)  | Current | New (latest commit) | Improvement |
-|-|-|-|
**16K**| 8.8s | 5.5s | 1.6x |
**100K** | 85.2s | 28.0s | 3x |
**1M** | 792s | 276s | 2.8x |

---

## Test Platform
* OpenJDK Runtime Environment Temurin-11.0.26+4 (build 11.0.26+4)
* GeoServer JVM flags: `-Xmx9G`
* Macbook Pro M3 Max
* 1TB SSD, APFS

## Data Directories:
* [large_shapefiles.tar.bz2](https://www.dropbox.com/scl/fi/ip3x82pdzui1rxnvwsqmb/large_shapefiles.tar.bz2?rlkey=7uk600hlcp9z6tdbsdka86dpl&st=bkz2knvy&dl=1):
  * 1K workspaces
  * 16K layers
* [100K_layers.tar.bz2](https://s3.us-east-1.amazonaws.com/geoserver.datadirectories/100K_layers.tar.bz2) (7.5MB)
  * 100 workspaces
  * 100K layers
* [1M_layers.tar.bz2](https://s3.us-east-1.amazonaws.com/geoserver.datadirectories/1M_layers.tar.bz2) (63.9MB)
  * 1K workspaces
  * 1M layers

## Test Scenarios:

Time to load the Catalog with `DEFAULT_LOGGING` logging profile. Timing taken from log entry `Read Catalog in XXX s, ...`

* **Current**:
  * `DefaultGeoServerLoader`
  * Strategy: 
    * Loads file contents in parallel using fixed thread pool, and adds them to a blocking queue.
    * Parses to `CatalogInfo` and adds them to the `Catalog` in the main thread (queue consumer).
    * This is **safe** but **slow**(ish) in terms of potential deadlocks due to GeoServerExtensions being called during the loading process.
* **New, queued adds**:
  * `DataDirectoryGeoServerLoader`
  * This is the current status of the community module, observing the behaviour mentioned by Andrea, at ec7a87ac4ec649b62045778ac6ebcf8a7b351666
  * Strategy:
    * Loads file contents in work-stealing thread pool, and adds them to a blocking queue.
    * Parses to `CatalogInfo` and adds them to the `Catalog` in the main thread (queue consumer).
    * Potentially offloads I/O (file reads) to additional `ForkJoinPool` threads to prevent blocking on the worker threads.
    * This is **safe** but **slow**(ish) in terms of potential deadlocks due to GeoServerExtensions being called during the loading process.
* **New, parallel parsing and adding**:
  * `DataDirectoryGeoServerLoader`
  * This is how the community module behaved before [[GEOS-11758] datadir-catalog-loader: Improve reliability and code quality](https://github.com/geoserver/geoserver/pull/8385)
  * Strategy:
    * Loads, parses, and adds `CatalogInfo`s to the catalog in work-stealing thread pool.
    * Potentially offloads I/O (file reads) to additional `ForkJoinPool` threads to prevent blocking on the worker threads.
    * This is **fast** but **unsafe**, for it produced **deadlocks** on some data directories when:
      * `XStreamPersister` ended up calling `GeoServerExtensions.bean()/extensions()` concurrently during startup. Spring holds a lock in the main thread for the singleton beans registry. `GeoServerExtensions` forces Spring resolving singletons outside its original plan, since it doesn't follow spring dependency injection practices, bypassing the container’s controlled lifecycle and dependency resolution mechanisms.
      * GeoTools factory lookups are performed concurrently when they're not yet initialized.
    * Workarounds were implemented, like forcing initializing the CRS subsystem, but they turned out not being enough.
    * Most importantly, [GEOS-11758] was executed largely because despite having eliminated the deadlocks in all the known data directories used, still couldn't be sure they wouldn't happen on other edge cases.
* **New, parallel parsing, sequential adds**:
  * `DataDirectoryGeoServerLoader`
  * This is the winner approach, **fast and safe**.
  * Strategy:
    * Loads **and** parses `CatalogInfo`s to the catalog in work-stealing thread pool.
    * Potentially offloads I/O (file reads) to additional `ForkJoinPool` threads to prevent blocking on the worker threads.
    * The `XStreamPersister` **does not resolve ResolvingProxy** references.
    * There are safe and unsafe `CatalogInfo` types in terms of when to resolve proxy references and threading:
      * Safe types are added to the catalog while in the worker threads (workspaces, namespaces, and styles)
      * Unsafe types (stores, resources, and published infos) are added to a queue once parsed, and their proxies resolved in the main (consumer) thread. These ones have the potential of resulting in calls to `GeoServerExtensions.bean/extensions()` and/or `ResourcePool.getDataStore()/getFeatureType()` while `Catalog.add()` validates them.
  

## Test Results

Data dir   | Current | New, queued adds | New, parallel parsing and adding | New, parallel parsing, sequential adds |
---|---|---|---|---|
**large_shapefiles**| 8.8s | 7.5s | 5.4s | 5.5s |
**100K_layers** | 1.42m | 48.3s | 29.9s | 28.0s |
**1M_layers** | 13.2m | *N/A* | *N/A* | 4.6m |

> SMB (or NFS for that matter), over a slow connection for this amount of small files just doesn't make sense. I couldn't find the time to run the tests against a 10Gbps network. But the **large_shapefiles** test scenario on SMB/1Gbps results in `19 secs` for the new loader vs. `45 secs` for the current one.

---

# Startup logs per scenario:

## large_shapefiles:

* **Current**

  ```
  10 Mar 16:23:38 CONFIG [org.geoserver] - Read Catalog in 8.828 s: workspaces: 1,008, namespaces: 1,008, styles: 27, stores: 2,011, resources: 16,025, layers: 16,025, layer groups: 4.
  SEVERE: GeoServer startup complete in 13.0s
  ```

* **New, queued adds**:

  ```
  10 Mar 16:25:52 CONFIG [org.geoserver] - Read Catalog in 7.563 s: workspaces: 1,008, namespaces: 1,008, styles: 27, stores: 2,011, resources: 16,025, layers: 16,025, layer groups: 4.
  SEVERE: GeoServer startup complete in 11.531s
  ```

* **New, parallel parsing and adding**:

  ```
  10 Mar 16:24:42 CONFIG [org.geoserver] - Read Catalog in 5.472 s: workspaces: 1,008, namespaces: 1,008, styles: 27, stores: 2,011, resources: 16,025, layers: 16,025, layer groups: 4.
  SEVERE: GeoServer startup complete in 9.325s
  ```

* **New, parallel parsing, sequential adds**:

  ```
  10 Mar 19:14:29 CONFIG [org.geoserver] - Read Catalog in 5.543 s: workspaces: 1,008, namespaces: 1,008, styles: 27, stores: 2,011, resources: 16,025, layers: 16,025, layer groups: 4.
  SEVERE: GeoServer startup complete in 9.396s
  ```

## 100K_layers:

* **Current**

  ```
  10 Mar 20:13:06 CONFIG [org.geoserver] - Read Catalog in 1.425 min: workspaces: 107, namespaces: 107, styles: 21, stores: 100,009, resources: 100,019, layers: 100,019, layer groups: 3.
  SEVERE: GeoServer startup complete in 89.357s
  ```

* **New, queued adds**:

  ```
  10 Mar 16:09:50 CONFIG [org.geoserver] - Read Catalog in 48.34 s: workspaces: 107, namespaces: 107, styles: 21, stores: 100,009, resources: 100,019, layers: 100,019, layer groups: 3.
  SEVERE: GeoServer startup complete in 55.616s
  ```

* **New, parallel parsing and adding**:

  ```
  10 Mar 16:17:06 CONFIG [org.geoserver] - Read Catalog in 28.94 s: workspaces: 107, namespaces: 107, styles: 21, stores: 100,009, resources: 100,019, layers: 100,019, layer groups: 3.
  SEVERE: GeoServer startup complete in 35.977s
  ```

* **New, parallel parsing, sequential adds**:

  ```
  10 Mar 19:12:55 CONFIG [org.geoserver] - Read Catalog in 28.05 s: workspaces: 107, namespaces: 107, styles: 21, stores: 100,009, resources: 100,019, layers: 100,019, layer groups: 3.
  SEVERE: GeoServer startup complete in 34.732s
  ```


## 1M_layers:

* **Current**

  ```
  10 Mar 21:02:58 CONFIG [org.geoserver] - Read Catalog in 13.27 min: workspaces: 1,007, namespaces: 1,007, styles: 21, stores: 1,000,009, resources: 1,000,019, layers: 1,000,019, layer groups: 3.
  SEVERE: GeoServer startup complete in 805.771s
  ```

* **New, parallel parsing, sequential adds**:

  ```
  10 Mar 21:09:14 CONFIG [org.geoserver] - Read Catalog in 4.681 min: workspaces: 1,007, namespaces: 1,007, styles: 21, stores: 1,000,009, resources: 1,000,019, layers: 1,000,019, layer groups: 3.
  SEVERE: GeoServer startup complete in 287.678s  
  ```

aaime commented:
Looks like all the builds are bricked:

![image](https://github.com/user-attachments/assets/7e76ad08-3829-4c4b-8f6b-3180f48e7140)

groldan commented:
My bad, I was missing [this](https://github.com/geoserver/geoserver/compare/a9b98fe39a1c52bf31d9dc92ba8124634e3ca357..8ed4c5d0a1d22101bfa8ce9d9ced070fb83014eb)
Maybe that calls for separate issue to be examined across all the geotools factory lookups, but I'm too drained out right now as to pursue it.
In any case, other than the CRS lookups, I've made sure no factory lookup will be performed by serializing calls to Catalog.add()

aaime commented:
Reminder to add a note to upgrade.rst explaining how to turn off the loader.

aaime commented:
Reminder to add a note to upgrade.rst explaining how to turn off the loader, just in case.

groldan commented:
> Reminder to add a note to upgrade.rst explaining how to turn off the loader, just in case.

[done](e8e5187d030f8819cb50f528416ddacbf898eb54)

jodygarnett reviewed (changes requested):
Some documentation feedback.

aaime commented:
Tested the latest drop against the 16k layers, it's still no faster than the normal catalog for me.

Also, I've found a case where the catalog is not being correctly loaded, I get this error:

```
14 mar 19:03:24 ERROR  [config.datadir] - Failed to load layer group LayerGroupInfoImpl--3cd36b64:16ef62e2f71:-7ff6
org.springframework.beans.factory.BeanCurrentlyInCreationException: Error creating bean with name 'localWorkspaceCatalog': Requested bean is currently in creation: Is there an unresolvable circular reference?
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.beforeSingletonCreation(DefaultSingletonBeanRegistry.java:355)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:227)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:334)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:209)
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1171)
	at org.geoserver.platform.GeoServerExtensions.getBean(GeoServerExtensions.java:249)
	at org.geoserver.platform.GeoServerExtensions.bean(GeoServerExtensions.java:293)
```

The net result is that GeoServer starts anyways, but only one of the layer groups is there (the only one that is not using style groups). The data directory is here: https://github.com/geosolutions-it/openmaptiles

Also, small thing, but I'm wondering if an approach similar to this (rough test mind) wouldn't be more efficient rather than polling the queue one item at a time:

```java
 private void consumeQueue(ForkJoinPool executor) {
        while (!executor.isShutdown()) {
            List<CatalogInfo> infos = new ArrayList<>();
            while (true) {

                sourceQueue.drainTo(infos);
                for (CatalogInfo info : infos) {
                    if (info == TERMINAL_TOKEN) {
                        return;
                    }
                    addToCatalog(info);
                }
                infos.clear();
            }
        }
    }
```

groldan commented:
There's some difference between running the web-app's Start class from eclipse and running e.g. `mvn -o -Pmbtiles-store,mbstyle jetty:run`. The former starts up with no problem, the later shows the error you mention. Looking into it.

aaime commented:
Probably some class path order difference? In my case it fails starting
from IntelliJ

Il ven 14 mar 2025, 21:29 Gabriel Roldan ***@***.***> ha
scritto:

> There's some difference between running the web-app's Start class from
> eclipse and running e.g. mvn -o -Pmbtiles-store,mbstyle jetty:run. The
> former starts up with no problem, the later shows the error you mention.
> Looking into it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/geoserver/geoserver/pull/8370#issuecomment-2725685081>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACPOOMMD5UZGFZCXLRCUVT2UM32RAVCNFSM6AAAAABYNLEAG6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRVGY4DKMBYGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: groldan]*groldan* left a comment (geoserver/geoserver#8370)
> <https://github.com/geoserver/geoserver/pull/8370#issuecomment-2725685081>
>
> There's some difference between running the web-app's Start class from
> eclipse and running e.g. mvn -o -Pmbtiles-store,mbstyle jetty:run. The
> former starts up with no problem, the later shows the error you mention.
> Looking into it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/geoserver/geoserver/pull/8370#issuecomment-2725685081>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACPOOMMD5UZGFZCXLRCUVT2UM32RAVCNFSM6AAAAABYNLEAG6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRVGY4DKMBYGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

aaime commented:
While looking into startup performance I've cooked up this one:
https://github.com/geoserver/geoserver/pull/8407

It brings the startup time from 120 seconds down to 20 seconds, with both the current and the new data dir loader. 
I still cannot confirm the speedup of the new data loader...

Also wondering how your tests weren't affected by the CRS lookup part of the above PR (to feel the second, you need the OGC API DGGS module loaded, it has a particularly slow "DataStoreFactory.isAvailable()" method as it needs to reach out to an external python process).

groldan commented on doc/en/user/source/datadirectory/setting.rst:
```diff
@@ -195,3 +195,53 @@ Java system property:
+
+* **Parallel Processing**: Both I/O calls and XML parsing of catalog and configuration files are parallelized
+* **Efficient Directory Traversal**: Makes a single pass over the ``workspaces`` directory tree, loading most catalog and configuration files in one pass
+* **Network Performance**: Particularly beneficial for deployments using network filesystems like NFS, which are typically slow when serving many small files
```
@jodygarnett let me know if the changes in c11752e95d4ff2fc88d931cb43d59069be1f3e30 are good enough

groldan commented:
@aaime the errors loading the openmaptiles datadir are fixed by these two:
* fdf9fad0f4d9aca9401901d052b0169fd299a6d0: Fix potential race conditions by ensuring `LayerGroupHelper` uses explicit catalog
* 75c5ac811995c92247af9d985e10dfb38d6aed0c: Refactor catalog proxy resolution into dedicated utility class

groldan commented:
I've conducted several tests with the following test data directories, with and without verbose logging, on NFS and a fast NVMe disk:
*[large_shapefiles.tar.bz2](https://www.dropbox.com/scl/fi/ip3x82pdzui1rxnvwsqmb/large_shapefiles.tar.bz2?rlkey=7uk600hlcp9z6tdbsdka86dpl&st=bkz2knvy&dl=1):
*[100K_layers.tar.bz2](https://s3.us-east-1.amazonaws.com/geoserver.datadirectories/100K_layers.tar.bz2)
*[1M_layers.tar.bz2](https://s3.us-east-1.amazonaws.com/geoserver.datadirectories/1M_layers.tar.bz2)

<img width="805" alt="image" src="https://github.com/user-attachments/assets/dd268d5a-4f36-43e9-a513-615124a23f1a" />

The current loader has indeed improved a lot since years ago, the new loader is still worth it IMO. For example, over NFS the limiting factor is I/O throughput and NFS struggles with many small files, but time ago the difference was much bigger than 3.6x.

petersmythe commented:
@jodygarnett asked for feedback on this implementation, can we fast-track the review and merge?

https://discourse.osgeo.org/t/testing-ahead-of-2-27-0-release/145727/3

aaime commented:
Gabriel asked the PSC for some tests, but testing so far is just me and him. If we cannot get more testing out of the PSC then I guess we should merge and let users do their thing.

jodygarnett commented:
I tested and while it works fine I do not really have an interesting data directory to stress the system.

I would like to recommend merging for public feedback, and if it does not go well we can default to having this loader "off" for release.

aaime commented:
It does not necessarily needs a big data directory, a rich combo of extensions (along with their configuration and startup behavior) is also an important test.

jodygarnett commented:
okay retesting with some reference data directories:

* [x] training (ysld,control-flow,ogcapi-features,wps) - all workspaces fine
* [x] demo (ysld,control-flow,ogcapi-features,wps, cog) - no workspace data dir, results ``Service Capabilities`` being listed on welcome page which is a worry not sure why
* [x] standard (ysld,control-flow,ogcapi-features,wps, cog) - all workspaces fine
* [x] default (ysld,control-flow,ogcapi-features,wps, cog) - no workspace data dir fine

aaime commented:
I guess we're not going to get any more feedback on this one eh? If so I'd say, let's merge and ask users.

jodygarnett commented:
Agreed, I will merge and ask for new feedback specifically on startup.

Thus far we are getting feedback on CSP (expected) GeoFence
